keps/sig-architecture/2527-clarify-status-observations-vs-rbac/README.md

```diff
@@ -0,0 +1,687 @@
+
+This KEP further proposes to add guidance for APIs that want to implement an
+"allocation" or "binding" pattern which requires trusted ACK.
+
+## Motivation
+
+As an API reviewer, I have seen many different patterns in use.  I have shot
+down APIs that run afoul of the from-observation test, and forced the
+implementations to be more complicated.  I no longer find this to be useful to
+our APIs, and in fact I find it to be a detriment.
+
+I suspect that several APIs would have come out differently if not for this.
+
+### Goals
+
+* Clarify or remove the from-observation rule to better match reality.
+* Provide guidance for APIs on how to save controller results.
+
+### Non-Goals
+
+* Retroactively apply this pattern to change existing GA APIs.
+* Necessarily apply this pattern to change pre-GA APIs (though they may choose
+  to follow it).
+* Provide arbitrary storage for controller state.
+* Provide distinct access control for subsets of spec or status.
+
+## Proposal
+
+This KEP proposes to:
+
+1) Get rid of the "if it got erased" litmus test for status fields
+2) Acknowledge that spec/status are a useful and meaningful split for access control
+3) Document one or more patterns for APIs that need a trusted acknowledgement
+   of the spec
+
+Depending on feedback, there amay be other approaches to solving the same
+problems.
+
+### Risks and Mitigations
+
+The current guidance exists for a reason.  It does encourage a sort of
+idealogical cleanliness.  However, it is not universally adhered to and it
+neglects the reality of these request/acknowledge APIs.  On the whole, this KEP
+posits that the current guidance is a net negative.
+
+## Design Details
+
+<<[UNRESOLVED ideas to debate]>>
+
+### Option 1: Loosen status
+
+Remove the idea that status fields must be from observation.  Allow controllers
+to write values to status that represent allocations or acknowledged requests.
+Encourage this pattern.  Document it.
+
+#### Examples
+
+Some of these are variants of existing APIs and some are hypothetical.
+
+1) User creates a Service and sets `Service.spec.type = LoadBalancer`.  The
+cloud controller sets `Service.status.loadBalancer.ingress.ip` in response.  If
+the user sets `Service.spec.loadBalancerIP` to a specific value, the cloud
+controller will either successfully use that IP and and set it as before (ACK),
+or it will not set any value in `Service.status.loadBalancer.ingress` (NAK).
+
+2) Given a Pod, user patches `Pod.spec.containers[0].resources.requests[cpu]`
+to a new value.  Kubelet sees this as a request and, if viable, sets
+`Pod.status.containerStatuses[0].resources.requests[cpu]` to the same value.
+
+3) User creates a `Bucket`, setting `Bucket.spec.instance = artifacts-prod`.
+The bucket controller verifies that the namespace is allowed to use the
+artifacts-prod bucket and, if so, sets `Bucket.status.instance` to the same
+value.
+
+#### Tradeoffs
+
+Pro: easy and compatible
+
+Con: dilutes the meaning of status
```

Comment:
This smacks of the "idealogically pure" comment above. I think that the reason for this rule is to ensure that if the status is lost somehow, we don't lose track of resources in the underlying platform the controller is managing. For example, we don't sever the connection between some cloud resource and the k8s cluster, and therefore basically leak the cloud resource.

I think that this is a nice principle, but I don't think it needs to be a rule rather, it can be a "best practice". Essentially we don't limit the design of our controllers, but we just document for controller authors that they should consider what happens if the k8s resource is destroyed without the controller having a chance to clean up. They should consider how they would ensure any leaked resources are identifiable out-of-band and removable.

Reply:
(in other words, I like this option if we add some docs around that)

Reply:
Added words